### PR TITLE
Scope lake tributaries

### PR DESCRIPTION
**Pull request by msleckman**
This PR scopes the tributaries of our focal lakes using `nhdplusTools::get_UT()`.

In our pipeline so far, we have been included **all** flowlines belonging to the huc08 basin boundaries of our lakes. This PR narrows the flowlines only the upstream tributaries of the lake. The gauges along these flowlines are expected to affect the lake conditions directly. 

#### Additions:
Adds target `p2_lake_tributaries` and `p2_lake_tributaries_cat` and a new function in `scope_lake_tributaries.R`. 

#### Process points:
* Since not all lakes have a lake ID and therefore no record of central outlet from which we can navigate upstream (unsuccessful attempt with `nhdplusTools::get_wb_outlet()` ) , the`scope_lake_tributaries()` custom function filtered to the flowline reaches within or touching our selected lake polygons (buffered) and pulls flowlines that are upstream from the lake. 

* Chose to buffer the lake polygons in order to capture the final reach that may not intersect lake polygon

* Chunked the upstream comids retrieved with `get_UT()` to be able fetch all flowlines without timeout errors

#### Outcome map: 

```
mapview(p2_lake_tributaries)+
#  mapview(comid_in_lake, color = 'red')+
  mapview(p2_saline_lakes_sf %>%
            st_buffer(dist= 10000), col.regions = NA,color = 'blue')+
  mapview(p2_saline_lakes_sf, col.regions = 'blue')+
  mapview(p1_lake_flowlines_huc8_sf, color = 'black')
```

**Lingering issues:**
* Obtaining s2 message when running `scope_lake_tributaries()` - stemming from `get_nhdplus()`
```
Spherical geometry (s2) switched on
Spherical geometry (s2) switched off
```

* Occasionally getting pipeline build failure stemming from `p1_lake_flowlines_huc8_sf`
```
Error in `tar_throw_run()`:
! callr subprocess failed: could not load dependencies of target p2_lake_tributaries_cat. Input must be a vector, not a <sfc_LINESTRING/sfc> object.
```
bandaid solution - running `tar_invalidate(p1_lake_flowlines_huc8_sf)` `before tar_make()`

Please review this PR but testing that the pipeline runs for you and producing the desired output.

### CODE REVIEW

**Review by amart90 — COMMENTED** [review]
Hi Margaux. The pipeline worked for me and ran in 14.28 minutes (I already had the data downloaded).
- I did not get the spherical geopetry warning or the tar_throw_run error you were having trouble with
- The rmd generates a map that looks like [this](https://user-images.githubusercontent.com/25806255/180457470-82833e0a-9888-452a-866a-adb6eefa8c2d.png), which looks right to me.
- It might be nice to have a binary variable in the flowlines data that describes wheter a flowline is in the lake katchment or not. It might help the viz to have flowlines in the catchment a more bold color than those not in the catchment, but still in the same subbasin.


#### Projection
[Disclaimer: The following comment comes from someone who is not a geographer. Someone with more insight into the topic is encouraged to add/correct any inaccuracies or oversights]... but, we should maybe touch on projections a little bit. Everything now is now being projected to the specified_crs of [EPSG: 4326](https://epsg.io/4326), which is [essentially unprojected](https://www.sharpgis.net/post/Why-EPSG4326-is-usually-the-wrong-e2809cprojectione2809d) and, especially if we turn off [spherical geometry](https://r-spatial.github.io/sf/articles/sf7.html) (see comment on [line 15 of 2_process/src/scope_lake_tributaries.R](https://github.com/USGS-VIZLAB/saline-lakes/pull/16/files#diff-47985969152656d3c332b6133b520ed416581bf2b02a0956b5fada2b4ccb8857R15), potentially going to lead to some inaccuracy. What makes this tricky is that leaflet requires things to be in EPSG 4326, from what I can tell. So, if we are going to send them shapefiles or use any of our products for analysis or measurement at all, I suggest that we make sure that we project the data into a different coordinate system, like [EPSG 5070](https://epsg.io/5070) and we can just project it back to 4326 just for plotting in leaflet.

**Comment by amart90** [comment]
>   _I think I am understanding what you are saying here, but please correct me if I am wrong_ - I think we could definitely  keep just one flowlines target for the entire basin in the huc08 areas of interest and have a binary col that distinguishes whether the given reach is within or outside the immediate lake catchment (or that it is a lake trib). This way we don't have two objects `p1_flowlines_` and `p2_lake_tributaries` and instead have a single target object (named `p2_lake_flowlines` ? ). A simple spatial join/filter might work with this.

@msleckman, yeah that's exactly what I was thinking. 

I think we can probably merge this PR and address these different _issues_ in the next PR.

**Review by amart90 — APPROVED** [review]

**Comment by padilla410** [comment]
**DRAFT**

Howdy Margaux,

I was able to successfully build the pipeline from scratch and to re-run it without running into unexpected rebuilds. I also confirmed that `p2_lake_tributaries`, `p2_lake_tributaries_cat`,  and `scope_lake_tributaries` are the only new entries in `tar_meta`.

I was not able to build the following outcome maps:
```r
mapview(p2_lake_tributaries)+
  mapview(comid_in_lake, color = 'red')+
  mapview(p2_saline_lakes_sf %>%
            st_buffer(dist= 10000), col.regions = NA,color = 'blue')+
  mapview(p2_saline_lakes_sf, col.regions = 'blue')+
  mapview(p1_lake_flowlines_huc8_sf, color = 'black')
```
because I do not have `comid_in_lake` in my environment nor is it a target. However, I could see pretty dramatic differences in the `PartnerMeeting.Rmd` output:

from `main` branch:
image


from `scope_lake_tributaries` branch:
image

**Review by padilla410 — APPROVED** [review]

**Comment by padilla410** [comment]
Ok so that was all a very long winded way of saying I approve under two conditions:
* we do the doc updates
* we can confirm that understanding I outlined above is correct

File: 2_process/src/scope_lake_tributaries.R [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+                                 realization = 'flowline',
+                                 stream_order = NULL){
+
+  #'@description subset flowlines network to get only tributaries upstream of lakes or catchments of lake Upstream tributaries
+  #'@paramfline_network sf dataframe of flowlines from nhdplustools 
+  #'@param lakes_sf sf dataframe of focal lake polygons
+  #'@param buffer_dist buffer distance for lakes to ensure we capture all flowlines relevant to lake
+  #'@param realization either flowline or catchment. Must input only 1, unlike with this param in get_nhdplus() 
+  #'@param stream_order stream order level. Defaults See get_nhdplus() for further details
+  #'@value nhdplus sf dataframe of flowlines or catchments of reaches Upstream of focal lakes
```
[the remark this answers is not in the document]
Comment (padilla410) [reply 2_process/src/scope_lake_tributaries.R]: I recognize this request is a bit of a pain. After I read all of your other comments, we could merge this PR and then open a new one for the doc changes. I should be able to have time to help you with the translation.

**Comment by padilla410** [comment]
Ok, we are good to go, Margaux. Let's open up the doc thing as an issue and merge this PR.

**Pull request merged**